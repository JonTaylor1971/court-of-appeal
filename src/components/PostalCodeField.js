import React, { Component } from 'react';

class PostalCodeField extends Component {

    constructor(props) {
        super(props);

        this.state = {
            className: 'form-field',
            title: ''
        };
        this.handleFieldChange = this.handleFieldChange.bind(this);
        this.validate = this.validate.bind(this);
        this.clearClass = this.clearClass.bind(this);
    }

    render() {
        let autofocus = this.props.autofocus ? true : false;
        return (<input
            id={this.props.id}
            name={this.props.name}
            autoFocus={autofocus}
            ref={this.props.ref}
            value={this.props.value || ''}
            onChange={this.handleFieldChange}
            onKeyPress={this.handleKeyPress}
            maxLength={7}
            size={30}
            title={this.state.title}
            className={this.state.className}
            onBlur={this.validate}
            onFocus={this.clearClass}
            placeholder="Valid Canadian postal code"
        />);
    }

    handleFieldChange(e) {
        let value = e.target.value;
        value = value.toUpperCase().replace(/[^\dA-Z ]/g,"");
        value = value.replace(/[ ]+/g," ");
        e.target.value =  value;
        this.props.handleFieldChange(e);
    }

    /**
     * If, when the field loses focus, it's not empty or a proper postal code, give the field an error color, and give
     * an error title message so they can see what's up when they hover.
     */
    validate() {
        if (this.props.value.match(/^[a-zA-Z]\d[a-zA-Z] \d[A-Z]\d$/) != null || !this.props.value) {
            this.setState({className: "form-field", title: ""});
        } else {
            this.setState({className: "form-field invalid-field", title: "ERROR: Invalid Postal Code"});
        }
    }

    clearClass() {
        this.setState({className: "form-field", title: ""});
    }


}
export default PostalCodeField;
