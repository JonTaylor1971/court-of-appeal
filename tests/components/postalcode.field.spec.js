require('../support/fake.dom');
import React from 'react';
import PostalCodeField from '../../src/components/PostalCodeField';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

test('should only allow valid Canadian postal code characters', ()=> {

    let value = '';
    let handleFieldChange = (e) => { value = e.target.value; };

    const postalCode = renderer.create(
        <PostalCodeField id="number-field" value={value}
                    handleFieldChange={handleFieldChange.bind(this)}/>,
    );
    let tree = postalCode.toJSON();
    expect(tree).toMatchSnapshot();

    let e = { target: { value: "42"}};

    let pc = postalCode.getInstance();

    pc.handleFieldChange(e);
    expect(value).toMatch('42');

    e = { target: { value: "bah" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('BAH');

    e = { target: { value: "heyBabaReeBear 123" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('HEYBABA');

    e = { target: { value: "V0N 1J22" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('V0N 1J2');

    e = { target: { value: "V0N   2" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('V0N 2');

});
