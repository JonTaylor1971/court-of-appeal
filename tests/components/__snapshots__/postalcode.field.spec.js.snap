// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should only allow valid Canadian postal code characters 1`] = `
<div>
  <input
    className="form-field"
    maxLength={7}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Valid Canadian postal code"
    readOnly={undefined}
    size={30}
    title=""
    value=""
  />
  <div
    className="error-message"
  >
    
  </div>
</div>
`;
